pkg/v2/tkr/webhook/infra-machine/fieldsetter/fieldsetter.go

```diff
@@ -0,0 +1,107 @@
+	"gopkg.in/yaml.v2"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"sigs.k8s.io/controller-runtime/pkg/webhook/admission"
+)
+
+const (
+	osImageRefAnnotationKey = "run.tanzu.vmware.com/os-image-ref"
+)
+
+// FieldSetter mutates <infra>Machine
+type FieldSetter struct {
+	decoder      *admission.Decoder
+	Log          logr.Logger
+	FieldPathMap map[string]string
+}
+
+// Handle will handle the infra admission request
+func (fs *FieldSetter) Handle(ctx context.Context, req admission.Request) admission.Response { // nolint:gocritic //req parameter is not pointer type as per standard signature of webhook callback handler
+	fs.Log.Info("Received the infra machine admission request")
+	infraMachine := &unstructured.Unstructured{}
+	err := fs.decoder.Decode(req, infraMachine)
+	if err != nil {
+		fs.Log.Error(err, "Failed to decode infra machine admission request")
+		return admission.Errored(http.StatusBadRequest, err)
+	}
+
+	annotationValues, err := getOSImageRefAnnotation(infraMachine)
+	if err != nil {
+		fs.Log.Error(err, "Failed to get OSImageref annotation")
+		return admission.ValidationResponse(true, err.Error())
+	}
+	err = fs.setFields(infraMachine, annotationValues)
+	if err != nil {
+		fs.Log.Error(err, "Failed to set the fields using the annotations")
+		return admission.ValidationResponse(true, err.Error())
+	}
+
+	marshalledInfraMachine, err := json.Marshal(infraMachine)
+	if err != nil {
+		fs.Log.Error(err, "Failed to marshal infraMachine object")
+		return admission.Errored(http.StatusInternalServerError, err)
+	}
+	return admission.PatchResponseFromRaw(req.Object.Raw, marshalledInfraMachine)
+}
+
+func getOSImageRefAnnotation(infraMachine *unstructured.Unstructured) (map[string]interface{}, error) {
+	annotations := infraMachine.GetAnnotations()
+	if annotations == nil {
+		return nil, errors.New("os Image reference annotation is missing")
+	}
+
+	if _, exists := annotations[osImageRefAnnotationKey]; !exists {
+		return nil, errors.New("os Image reference annotation is missing")
+	}
+	annotationValues := make(map[string]interface{}, 1)
+	err := yaml.Unmarshal([]byte(annotations[osImageRefAnnotationKey]), &annotationValues)
+	if err != nil {
+		return nil, errors.New("failed to unmarshal os Image reference annotation value")
+	}
+	return annotationValues, nil
+}
+
+// SetFields sets the <Infra>Machine fields using the annotation values
+func (fs *FieldSetter) setFields(o *unstructured.Unstructured, annotationValues map[string]interface{}) error {
+	for field, value := range annotationValues {
+		path, exists := fs.FieldPathMap[field]
+		if !exists {
+			msg := fmt.Sprintf("os Image reference annotation value's field %q doesn't match with any entry in field path map configured", field)
+			fs.Log.Info(msg)
+			continue
+		}
+		fieldPath := strings.Split(path, ".")
+		// skip setting the field if the path doesn't exists
+		if _, exists, _ := unstructured.NestedFieldNoCopy(o.UnstructuredContent(), fieldPath...); !exists {
```

Comment:
Swallowing an error here (assigning to `_`). We should handle it instead.

Reply:
This is just safe check to know if the path exists. The actual error handling is handled while setting the field at L95.